hbase-client/src/main/java/org/apache/hadoop/hbase/client/trace/IpcClientSpanBuilder.java

```diff
@@ -0,0 +1,103 @@
+import static org.apache.hadoop.hbase.trace.HBaseSemanticAttributes.RPC_METHOD;
+import static org.apache.hadoop.hbase.trace.HBaseSemanticAttributes.RPC_SERVICE;
+import static org.apache.hadoop.hbase.trace.HBaseSemanticAttributes.RPC_SYSTEM;
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanBuilder;
+import io.opentelemetry.api.trace.SpanKind;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+import org.apache.hadoop.hbase.net.Address;
+import org.apache.hadoop.hbase.trace.HBaseSemanticAttributes.RpcSystem;
+import org.apache.hadoop.hbase.trace.TraceUtil;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.apache.hbase.thirdparty.com.google.protobuf.Descriptors;
+
+/**
+ * Construct {@link Span} instances originating from the client side of an IPC.
+ */
+@InterfaceAudience.Private
+public class IpcClientSpanBuilder implements Supplier<Span> {
+
+  private String name;
+  private final Map<AttributeKey<?>, Object> attributes = new HashMap<>();
+
+  @Override
+  public Span get() {
+    return build();
+  }
+
+  public IpcClientSpanBuilder setMethodDescriptor(final Descriptors.MethodDescriptor md) {
+    final String packageAndService = getRpcPackageAndService(md);
+    final String method = getRpcName(md);
+    this.name = packageAndService + "/" + method;
+    populateMethodDescriptorAttributes(attributes, md);
+    return this;
+  }
+
+  public IpcClientSpanBuilder setRemoteAddress(final Address remoteAddress) {
+    attributes.put(NET_PEER_NAME, remoteAddress.getHostName());
+    attributes.put(NET_PEER_PORT, (long) remoteAddress.getPort());
+    return this;
+  }
+
+  @SuppressWarnings("unchecked")
+  public Span build() {
+    final SpanBuilder builder = TraceUtil.getGlobalTracer()
+      .spanBuilder(name)
+      // TODO: what about clients embedded in Master/RegionServer/Gateways/&c?
+      .setSpanKind(SpanKind.CLIENT);
+    attributes.forEach((k, v) -> builder.setAttribute((AttributeKey<? super Object>) k, v));
+    return builder.startSpan();
+  }
+
+  /**
+   * Static utility method that performs the primary logic of this builder. It is visible to other
+   * classes in this package so that other builders can use this functionality as a mix-in.
+   * @param attributes the attributes map to be populated.
+   * @param md the source of the RPC attribute values.
+   */
+  static void populateMethodDescriptorAttributes(
+    final Map<AttributeKey<?>, Object> attributes,
+    final Descriptors.MethodDescriptor md
+  ) {
+    final String packageAndService = getRpcPackageAndService(md);
+    final String method = getRpcName(md);
+    attributes.put(RPC_SYSTEM, RpcSystem.HBASE_RPC.name());
+    attributes.put(RPC_SERVICE, packageAndService);
+    attributes.put(RPC_METHOD, method);
+  }
+
+  private static String getRpcPackageAndService(final Descriptors.MethodDescriptor md) {
```

Comment:
We could move this method to a util class and also let the IpcServerSpanBuilder to call this method too? So in the future if we want to change the implementation, we could fix them both at the same time.

Reply:
I'm generally against "util" classes. They indicate to me a lack of proper object/interface design and separation of concerns. I also prefer that test code reuse as little of the implementation code as is necessary -- and no more. The reason being that test code usually needs to test output behaviors. When test code and implementation share the same code for computing those behaviors, it nullifies the point.

Let me see what makes sense for these methods.

Reply:
I follow this strategy in the latest patch. `IpcServerSpanBuilder` has the details of how these tracing variables are extracted from the `MethodDescriptor` instance. The test uses constant values rather than looking them up from some instance. I think this makes for a more robust test.

Reply:
But IpcServerSpanBuilder is not for testing only? It is what we need to use at server side? I can not follow...

Reply:
I misunderstood you. The latest commit consolidates the implementation for both `IpcClientSpanBuilder` and `IpcServerSpanBuilder` into `IpcClientSpanBuilder`.